### ISSUE TITLE
Collapse bug

### ISSUE THREAD

**Issue by calintat** [homotopy-io/homotopy-rs#798]
I am afraid I found another bug in collapse.

Construct an endomorphism 4-cell on the braid:

![homotopy_io_export(1)](https://user-images.githubusercontent.com/16837293/189634170-52dd9945-e0e3-4dd1-aa07-86eca20e7765.svg) : ![homotopy_io_export(2)](https://user-images.githubusercontent.com/16837293/189634518-24868b8a-e254-4a38-91cb-4148e0eb9caa.svg) $\rightarrow$ ![homotopy_io_export(2)](https://user-images.githubusercontent.com/16837293/189634518-24868b8a-e254-4a38-91cb-4148e0eb9caa.svg)

Then go inside the source/target slice and attach a braid like this:

![homotopy_io_export(3)](https://user-images.githubusercontent.com/16837293/189634591-e6acf75f-f0ef-4b0a-90b7-4a5f03909877.svg)

Go into the projected view and try to contract the 3-cell and this homotopy:

![homotopy_io_export(4)](https://user-images.githubusercontent.com/16837293/189634747-db4b866f-3708-40af-a9fb-b07fe510dc89.svg)

This crashes with "Label inconsistency" 😢

**Comment by calintat** [comment]
Here is the action dump if that's useful: 
[homotopy_io_crash_action_dump(1).txt](https://github.com/homotopy-io/homotopy-rs/files/9547240/homotopy_io_crash_action_dump.1.txt)

**Comment by jamievicary** [comment]
Thanks for finding this. These are the final squeals as homotopy theory gives up its secrets to us. To be clear are you saying this is a panic, rather than simply a failure to contract. If so that's good as it is very surely fixable. Panics are easier to fix than mere incorrect behaviour.

An aside --- why is the endomorphism homotopy not represented with a vertex?

On phone today so haven't actually loaded this yet in the tool, sorry if I've misunderstood something.

**Comment by calintat** [comment]
> Thanks for finding this. These are the final squeals as homotopy theory gives up its secrets to us. To be clear are you saying this is a panic, rather than simply a failure to contract. If so that's good as it is very surely fixable. Panics are easier to fix than mere incorrect behaviour.

Yes, it's an explicit panic in the code, so something that shouldn't happen.

**Comment by calintat** [comment]
> An aside --- why is the endomorphism homotopy not represented with a vertex?

I was wondering about that too. The reason I think is that the homotopy is occluded by the identity in front of it.

**Comment by jamievicary** [comment]
Also bear in mind that since you use expansion to build the braid, there is a small chance this is actually a consequence of the known expansion bug.

**Comment by calintat** [comment]
This also panics if you replace the braid with a semi-braid and you don't need to do any interchangers (which is where the bug in expansion is) to construct that:

![homotopy_io_export(1)](https://user-images.githubusercontent.com/16837293/189644154-3f75657a-61a5-4691-b734-c7541d07f64c.svg)

**Comment by NickHu** [comment]
This panic is basically a debug assertion; when I wrote the code I thought it should never happen. If I merely disable it, it seems to do the right thing, and doesn't trip the well-formedness checker. It requires a bit more thought

**Comment by calintat** [comment]
I think that panic should never happen. It means we're trying to collapse a non-identity edge.

**Comment by NickHu** [comment]
> I think that panic should never happen. It means we're trying to collapse a non-identity edge.

This is only true if I implemented it (I guess in this case 'it' means `unify`) correctly

**Comment by calintat** [comment]
It's interesting that everything seems to work if you ignore that non-identity edge, the rest of the collapse is fine. So the equivalence classes are almost correct except for that edge.

**Comment by calintat** [comment]
> > I think that panic should never happen. It means we're trying to collapse a non-identity edge.
> 
> 
> 
> This is only true if I implemented it (I guess in this case 'it' means `unify`) correctly

I looked at unity earlier and I was a bit confused why you collapse one edge at a time, instead of just collapsing all edges together. Also is it really necessary to do this in place? Why can't we just reconstruct the graph like we used to.

**Comment by calintat** [comment]
I'm not saying it's wrong, it just seemed too complicated.

**Comment by NickHu** [comment]
The main reason to do it in-place is because now we would have to reconstruct the graph many times (one for each collapse subproblem, as dictated by the tree). So it seemed better this way from both a conceptual and performance standpoint.

**Comment by jamievicary** [comment]
By tree depth do you mean the way in which we currently prioritise edges, ie collapsing the deepest edges first.

**Comment by jamievicary** [comment]
By the way this is why we should have as many debug assertions as possible in the code.

**Comment by calintat** [comment]
> By the way this is why we should have as many debug assertions as possible in the code.

This is a debug assertion!

**Comment by NickHu** [comment]
This is not fixed by 9a87334f7b204b940088c4c3d35903fe84453836.

```
panicked at 'assertion failed: `(left == right)`
  left: `Some((2, BoundaryPath(Source, 1), [Singular(0), Singular(0)]))`,
 right: `Some((2, BoundaryPath(Target, 0), [Singular(0), Singular(0), Singular(0)]))`', homotopy-core/src/contraction.rs:455:29
```

**Comment by calintat** [comment]
@NickHu is this also fixed?

**Comment by NickHu** [comment]
Fixed by a02898f9629e12544f916c4d8c094b4b1fbb095a

I also tested on your new file Jamie and it works fine.

**Comment by jamievicary** [comment]
Can you summarize what the issue was?

**Comment by NickHu** [comment]
There were actually two issues:
1. I was only checking for refutations against an edge being collapsible locally within the collapse subproblem; now I'm just checking all the edges.
2. I was previously scanning through the graph, looking for collapsible edges, e.g. finding edges $E = x, y, z, \ldots$, and then collapsing $x, y, z, \ldots$ in turn (due to Rust borrowing and ownership). But of course, collapsing $x$ might actually make $y$ not collapsible, or even mean that there are new collapsible edges not contained in $E$ (at least, this is the behaviour I observed, I didn't think too hard about when this happens). Now it finds the fixed point of collapsing along collapsible edges, one edge at a time (while loop).

**Comment by jamievicary** [comment]
Thanks, is this all for a given edge depth? (In the sense that we first contract the deepest edges.)

**Comment by NickHu** [comment]
Yes, it still does the deepest edges first.

**Comment by jamievicary** [comment]
So you are saying that within a given set of edges of fixed depth, we cannot just collapse all the collapsible edges without re-checking collapsibility. That is surprising and potentially problematic -- in particular it means that result may depend on the collapse order which is currently arbitrary. Something to think about.

**Comment by calintat** [comment]
> There were actually two issues:
> 
> 1. I was only checking for refutations against an edge being collapsible locally within the collapse subproblem; now I'm just checking all the edges.
> 
> 2. I was previously scanning through the graph, looking for collapsible edges, e.g. finding edges $E = x, y, z, \ldots$, and then collapsing $x, y, z, \ldots$ in turn (due to Rust borrowing and ownership). But of course, collapsing $x$ might actually make $y$ not collapsible, or even mean that there are new collapsible edges not contained in $E$ (at least, this is the behaviour I observed, I didn't think too hard about when this happens). Now it finds the fixed point of collapsing along collapsible edges, one edge at a time (while loop).

This is interesting and makes perfect sense!

**Comment by NickHu** [comment]
Fixed by e9637e4078870f0d585b5ece60b1e31a8945e268
